File: selection/validate_suite.php
<?php
ini_set('memory_limit', '2048M');  // offline validator: heavy tries (e.g. 2000-word) exceed the 128M default
// Full-suite correctness validator (PHP). Runs solution.php against EVERY
// reference case and compares each result to the expected output. Run FROM the
// cell dir:  php ../../../selection/validate_suite.php
// Reuses harness.php's tree/list/canon logic so correctness judges the exact
// same shared inputs the measurement harness does.
// Prints one VALIDATE line to STDERR. Exit: 0 Accepted, 1 Wrong Answer,
// 3 Runtime Error, 2 setup error.

class TreeNode { public $val; public $left = null; public $right = null;
    function __construct($v = 0) { $this->val = $v; } }
class ListNode { public $val; public $next = null;
    function __construct($v = 0) { $this->val = $v; } }

function buildTree($a) {
    if (!is_array($a) || count($a) === 0 || $a[0] === null) return null;
    $root = new TreeNode($a[0]); $q = [$root]; $qi = 0; $i = 1; $n = count($a);
    while ($i < $n && $qi < count($q)) {
        $node = $q[$qi++];
        if ($i < $n) { $lv = $a[$i++]; if ($lv !== null) { $node->left  = new TreeNode($lv); $q[] = $node->left; } }
        if ($i < $n) { $rv = $a[$i++]; if ($rv !== null) { $node->right = new TreeNode($rv); $q[] = $node->right; } }
    }
    return $root;
}
function buildList($a) {
    if (!is_array($a) || count($a) === 0) return null;
    $head = new ListNode($a[0]); $c = $head;
    for ($i = 1; $i < count($a); $i++) { $c->next = new ListNode($a[$i]); $c = $c->next; }
    return $head;
}
function treeToArr($r) {
    $a = []; $q = [$r]; $qi = 0;
    while ($qi < count($q)) {
        $n = $q[$qi++];
        if ($n === null) { $a[] = null; }
        else { $a[] = $n->val; $q[] = $n->left; $q[] = $n->right; }
    }
    while (count($a) && end($a) === null) array_pop($a);
    return $a;
}
function listToArr($h) { $a = []; while ($h !== null) { $a[] = $h->val; $h = $h->next; } return $a; }
function canon($r) {
    if ($r instanceof ListNode) return json_encode(listToArr($r));
    if ($r instanceof TreeNode) return json_encode(treeToArr($r));
    $s = json_encode($r);
    return $s === false ? strval($r) : $s;
}
// design/trace: a null in expected marks a void op (LeetCode discards its
// return); compare strictly only at value-returning positions.
function seq_ok($actual, $expected) {
    if (!is_array($actual) || !is_array($expected) || count($actual) !== count($expected)) return false;
    for ($i = 0; $i < count($expected); $i++) {
        if ($expected[$i] === null) continue;
        if (canon($actual[$i]) !== canon($expected[$i])) return false;
    }
    return true;
}

$cell = getcwd();
$slug = basename($cell);
$ref  = $cell . '/../../../reference/leetcode';

$out = @json_decode(@file_get_contents($ref . '/outputs/' . $slug . '.json'), true);
$wl  = @json_decode(@file_get_contents($ref . '/workloads/' . $slug . '.json'), true);
if ($out === null || $wl === null) {
    fwrite(STDERR, "VALIDATE slug=$slug ERROR load: missing reference json\n"); exit(2);
}
$ep = strval($wl['entry_point'] ?? '');
$parts = explode('.', $ep); $method = end($parts);
$randomized = ($slug === 'random-pick-index');
// LeetCode accepts these answers in ANY order (special judge) -> multiset compare.
$UNORDERED = ['uncommon-words-from-two-sentences', 'remove-invalid-parentheses', 'restore-the-array-from-adjacent-pairs'];
function unordered_eq($a, $e) {
    if (!is_array($a) || !is_array($e)) return canon($a) === canon($e);
    $ka = array_map('canon', $a); sort($ka);
    $ke = array_map('canon', $e); sort($ke);
    return $ka === $ke;
}

// ---- load the solution (strip stray leading <?php, eval) ----------------
$src = @file_get_contents($cell . '/solution.php');
if ($src === false) { fwrite(STDERR, "VALIDATE slug=$slug ERROR load: no solution.php\n"); exit(2); }
$src = preg_replace('/^\xEF\xBB\xBF/', '', $src);
$src = preg_replace('/^\s*<\?php\b/', '', $src, 1);
$src = preg_replace('/^\s*<\?/', '', $src, 1);
$src = preg_replace('/\?' . '>\s*$/', '', $src, 1);
$before = get_declared_classes();
try { eval($src); }
catch (\Throwable $e) { fwrite(STDERR, "VALIDATE slug=$slug ERROR load: $e\n"); exit(2); }
$newClasses = array_values(array_diff(get_declared_classes(), $before));
$hasSol = class_exists('Solution');

function run_case($input, $method, $hasSol, $newClasses, $randomized) {
    if (isset($input['ops']) && isset($input['args'])) {
        $ops = $input['ops']; $args = $input['args'];
        $cls = null;
        if (class_exists($ops[0])) $cls = $ops[0];
        elseif (class_exists('Solution')) $cls = 'Solution';
        elseif (count($newClasses)) $cls = $newClasses[0];
        if ($cls === null) throw new \RuntimeException("no class {$ops[0]}");
        $nums = $randomized && isset($args[0][0]) ? $args[0][0] : null;
        $inst = new $cls(...$args[0]);
        $seq = [null];
        for ($i = 1; $i < count($ops); $i++) {
            $r = $inst->{$ops[$i]}(...$args[$i]);
            if ($randomized && $ops[$i] === 'pick' && is_int($r) && $nums !== null) $r = $nums[$r];
            $seq[] = $r;
        }
        return $seq;
    }
    $base = [];
    foreach (array_keys($input) as $k) {
        if ($k === 'root')     $base[] = buildTree($input[$k]);
        elseif ($k === 'head') $base[] = buildList($input[$k]);
        else                   $base[] = $input[$k];
    }
    return $hasSol ? (new Solution())->$method(...$base) : $method(...$base);
}

$__total = count($out['expected']);
foreach ($out['expected'] as $__i => $c) {
    $name = $c['name'];
    $input = $c['input'];
    $is_design = isset($input['ops']) && isset($input['args']);
    try {
        $actual = run_case($input, $method, $hasSol, $newClasses, $randomized);
    } catch (\Throwable $e) {
        fwrite(STDERR, "VALIDATE slug=$slug RE case=$name passed=$__i ncases=$__total " . get_class($e) . ": " . $e->getMessage() . "\n");
        exit(3);
    }
    $ok = $is_design ? seq_ok($actual, $c['output'])
         : (in_array($slug, $UNORDERED, true) ? unordered_eq($actual, $c['output']) : (canon($actual) === canon($c['output'])));
    if (!$ok) {
        fwrite(STDERR, "VALIDATE slug=$slug FAIL case=$name passed=$__i ncases=$__total "
            . "expected=" . substr(canon($c['output']), 0, 120) . " "
            . "actual="   . substr(canon($actual), 0, 120) . "\n");
        exit(1);
    }
}
fwrite(STDERR, "VALIDATE slug=$slug PASS ncases=$__total passed=$__total\n");
exit(0);

File: selection/validate_suite.py
#!/usr/bin/env python3
"""Full-suite correctness validator (Python).

Runs the cell's solution.py against EVERY reference case for the problem and
compares each result to the expected output. Unlike the old `make validate`
(which for LeetCode cells diffs stdout against /dev/null and so checks nothing),
this is a real correctness oracle: PASS only if every case matches.

Run FROM the cell dir (Energy-Languages/Python/leetcode/<slug>/):
    python3 ../../../selection/validate_suite.py

Reads the shared, language-independent cases from
reference/leetcode/{workloads,outputs}/<slug>.json — the same source the
measurement harness uses, so correctness and performance judge identical inputs.

Prints ONE line to stderr and exits:
    VALIDATE slug=<slug> PASS ncases=N                 -> exit 0  (Accepted)
    VALIDATE slug=<slug> FAIL case=<name> ...          -> exit 1  (Wrong Answer)
    VALIDATE slug=<slug> RE case=<name> ...            -> exit 3  (Runtime Error)
    VALIDATE slug=<slug> ERROR ...                     -> exit 2  (load/setup)
"""
import sys, os, json, copy, importlib.util
from collections import deque

sys.setrecursionlimit(1_000_000)  # match the measurement harness / LeetCode (deep recursive trees)

CELL = os.getcwd()
SLUG = os.path.basename(CELL)
REF = os.path.normpath(os.path.join(CELL, "..", "..", "..", "reference", "leetcode"))

# Problems where LeetCode accepts the answer in ANY order (special judge):
# compare as a multiset so a correct permutation is not a false Wrong Answer.
_UNORDERED = {"uncommon-words-from-two-sentences", "remove-invalid-parentheses",
              "restore-the-array-from-adjacent-pairs"}


# Node types LeetCode injects for tree/list problems (the solution references
# `TreeNode`/`ListNode` without defining them). Built from the shared level-order
# / array serialization; a node RESULT is canonicalised back the same way.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val; self.left = left; self.right = right


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val; self.next = next


def build_tree(vals):
    if not vals or vals[0] is None:
        return None
    root = TreeNode(vals[0]); q = deque([root]); i = 1
    while q and i < len(vals):
        n = q.popleft()
        if i < len(vals) and vals[i] is not None:
            n.left = TreeNode(vals[i]); q.append(n.left)
        i += 1
        if i < len(vals) and vals[i] is not None:
            n.right = TreeNode(vals[i]); q.append(n.right)
        i += 1
    return root


def build_list(vals):
    if not vals:
        return None
    head = ListNode(vals[0]); cur = head
    for v in vals[1:]:
        cur.next = ListNode(v); cur = cur.next
    return head


def _load():
    w = json.load(open(os.path.join(REF, "workloads", SLUG + ".json")))
    ep = (w.get("entry_point") or "").strip()
    method = ep.split(".")[-1].replace("()", "") if ep else None
    o = json.load(open(os.path.join(REF, "outputs", SLUG + ".json")))
    return method, o["expected"]


def _load_solution():
    spec = importlib.util.spec_from_file_location(
        "solution", os.path.join(CELL, "solution.py"))
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    # Provide TreeNode/ListNode the way LeetCode's judge does, unless the cell
    # ships its own. Attribute access is duck-typed, so canon works either way.
    if not hasattr(mod, "TreeNode"):
        mod.TreeNode = TreeNode
    if not hasattr(mod, "ListNode"):
        mod.ListNode = ListNode
    return mod


def _norm(v):
    """Normalise for comparison: TreeNode/ListNode -> array shape, tuples -> lists
    (JSON has no tuples), recurse. Nodes are detected duck-typed so a solution's
    own TreeNode/ListNode class canonicalises the same way."""
    if v is None:
        return None
    if hasattr(v, "val") and hasattr(v, "next"):            # ListNode
        out = []
        while v is not None:
            out.append(v.val); v = v.next
        return out
    if hasattr(v, "val") and (hasattr(v, "left") or hasattr(v, "right")):  # TreeNode
        out = []; q = deque([v])
        while q:
            n = q.popleft()
            if n is None:
                out.append(None); continue
            out.append(n.val); q.append(n.left); q.append(n.right)
        while out and out[-1] is None:
            out.pop()
        return out
    if isinstance(v, (tuple, list)):
        return [_norm(x) for x in v]
    return v


def _run_case(mod, method, case):
    """Execute one case, returning the actual result in the SAME shape as the
    stored expected output. Handles call-mode, design/trace-mode, and the
    randomized `random-pick-index` (validity: nums[pick] must equal target)."""
    inp = case["input"]
    # ---- design / trace mode: ops[0] is the class, the rest are method calls
    if isinstance(inp, dict) and "ops" in inp and "args" in inp:
        ops, args = inp["ops"], inp["args"]
        cls = getattr(mod, ops[0], None)
        if cls is None:
            # The trace's class name differs from what solution.py defines —
            # fall back to the single class the module exposes.
            classes = [v for v in vars(mod).values() if isinstance(v, type)
                       and v.__module__ == mod.__name__]
            cls = classes[0] if len(classes) == 1 else None
        if cls is None:
            raise RuntimeError(f"no class {ops[0]!r} in solution.py")
        randomized = (SLUG == "random-pick-index")
        nums = args[0][0] if randomized and args and args[0] else None
        obj, out = None, []
        for op, a in zip(ops, args):
            aa = copy.deepcopy(a)
            if obj is None:                       # first entry = constructor
                obj = cls(*aa)
                out.append(None)
            else:
                r = getattr(obj, op)(*aa)
                # random-pick returns a random valid index; the reference output
                # stores nums[index] (== target for any correct solution), so
                # canonicalise the same way instead of comparing raw indices.
                if randomized and op == "pick" and isinstance(r, int) and nums is not None:
                    r = nums[r]
                out.append(r)
        return _norm(out)
    # ---- call mode: input is {argname: value}; root/head build tree/list nodes
    target = mod.Solution() if hasattr(mod, "Solution") else mod
    fn = getattr(target, method)
    fargs = []
    for k, v in inp.items():
        if k == "root":
            fargs.append(build_tree(v))
        elif k == "head":
            fargs.append(build_list(v))
        else:
            fargs.append(copy.deepcopy(v))
    return _norm(fn(*fargs))


def _seq_ok(actual, expected):
    """design/trace: a null in expected marks a void op (LeetCode discards its
    return); compare strictly only at value-returning positions."""
    if not isinstance(actual, list) or not isinstance(expected, list) or len(actual) != len(expected):
        return False
    return all(e is None or _norm(a) == _norm(e) for a, e in zip(actual, expected))


def _unordered_eq(actual, expected):
    if not isinstance(actual, list) or not isinstance(expected, list):
        return actual == expected
    key = lambda xs: sorted(json.dumps(x, sort_keys=True) for x in xs)
    return key(actual) == key(expected)


def main():
    try:
        method, cases = _load()
        mod = _load_solution()
    except Exception as e:  # noqa: BLE001
        sys.stderr.write(f"VALIDATE slug={SLUG} ERROR load: {e}\n")
        sys.exit(2)

    for c in cases:
        name = c.get("name")
        try:
            actual = _run_case(mod, method, c)
        except Exception as e:  # noqa: BLE001
            sys.stderr.write(
                f"VALIDATE slug={SLUG} RE case={name} {type(e).__name__}: {e}\n")
            sys.exit(3)
        expected = _norm(c["output"])
        inp = c["input"]
        is_design = isinstance(inp, dict) and "ops" in inp and "args" in inp
        if is_design:
            ok = _seq_ok(actual, expected)
        elif SLUG in _UNORDERED:
            ok = _unordered_eq(actual, expected)
        else:
            ok = (actual == expected)
        if not ok:
            sys.stderr.write(
                f"VALIDATE slug={SLUG} FAIL case={name} "
                f"expected={json.dumps(expected)[:120]} "
                f"actual={json.dumps(actual, default=str)[:120]}\n")
            sys.exit(1)

    sys.stderr.write(f"VALIDATE slug={SLUG} PASS ncases={len(cases)}\n")
    sys.exit(0)


if __name__ == "__main__":
    main()
